Update readme-developer to Markdown, centralizing build instructions (#5704)
endless-sky/endless-sky: readme-developer.md

##### DEB-based distros:
   - g++
   - scons
   - libsdl2-dev
   - libpng-dev
   - libjpeg-dev
   - libgl1-mesa-dev
   - libglew-dev
   - libopenal-dev
   - libmad0-dev

##### RPM-based distros:
   - gcc-c++
   - scons
   - SDL2-devel
   - libpng-devel
   - libjpeg-turbo-devel
   - mesa-libGL-devel
   - glew-devel
   - openal-soft-devel
   - libmad-devel

Then, from the project root folder, simply type:

```
    $ scons
    $ ./endless-sky
```

The program will run, using the "data" and "images" folders that are found in the project root. For more Linux help, and to view other command-line flags, consult the `man` page (endless-sky.6), by running `man endless-sky` in the project directory.
The program also accepts a `--help` command line flag.

To compile and also run unit tests, the "test" target can be used:

```
  $ scons test
```

The produced test binary will be invoked with a reasonable set of arguments. To run any test benchmarks, you will need to invoke the test binary directly. Refer to the build pipeline definitions for the current required arguments.
More information on unit tests can be found in the project's "tests" directory.

## Windows

> Two methods of building the game on Windows are available: via the Code::Blocks IDE, and via Scons on a command line. Both options require a MinGW toolchain, and precompiled libraries provided below.

Reviewer: ```suggestion
Two methods of building the game on Windows are supported. The official build is done on the command-line with the Scons program, but the game may also be built using the Code::Blocks IDE. Both require a MinGW toolchain and the precompiled libraries linked below.
```

MinGW provides the tools used to compile the source code into the game's executable. For most cases, you should download the [MinGW-W64 Online Installer](https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/installer/mingw-w64-install.exe). When installing, use the following settings:

> - Version: `8.1.0`

Reviewer: ```suggestion
- Version: `8.1.0` (or greater; this document assumes 8.1.0)
```

- Architecture: `x86_64`
- Threads: `posix`
- Exception: `seh`

> - Build revision: `0`

Reviewer: ```suggestion
```
We shouldn't specify the revision number

> Endless Sky will also require precompiled libraries to build and run. Download them [here](http://endless-sky.github.io/win64-dev.zip), extract the zip, then move the `dev64` folder to `C:\` in order to build the game without further configuration.

Reviewer: ```suggestion
Endless Sky requires precompiled libraries to compile and play: [Download link](http://endless-sky.github.io/win64-dev.zip)

The zip can be extracted anywhere on your filesystem; to minimize additional configuration you can move the `dev64` folder to `C:\`.
```

### Building with CodeBlocks

If you want to build the game with a visual IDE, then you can use [CodeBlocks](https://codeblocks.org/downloads/26), the main supported IDE for Endless Sky on Windows. As you have already installed a compiler via MinGW, you should select a version without one.

After you install it, CodeBlocks will likely detect a "GNU GCC Compiler" from MinGW. Select this as your compiler, then take these steps to configure it:

- Go to Settings > Compiler > Global Compiler Settings > Toolchain Executables. Set the **Compiler's installation directory** to `C:\Program Files\mingw-w64\x86_64-8.1.0-posix-seh-rt_v6-rev0\mingw64\bin`, if you followed the default MinGW installation process above.
- Modify **Linker for static libs** from `ar.exe` to `gcc-ar.exe`.

> After this, the game should be ready to compile. Use CodeBlocks to open EndlessSky.workspace in the directory where you've cloned the game's code. This workspace consists of three projects: one for the majority of source code, one for the game binary, and one for the unit tests. As with most IDE-based projects, you must explicitly add new files to the respective project for them to be compiled.

Reviewer: ```suggestion
After this, the game should be ready to compile. Use CodeBlocks to open the "EndlessSky.workspace" file in the directory where you've cloned the game's code. This workspace consists of three projects: one for the majority of source code, one for the game binary, and one for the unit tests. As with most IDE-based projects, you must explicitly add new files to the respective project for them to be compiled.
```

With the workspace open, you can double-click a project in the left-hand menu to activate that project, binding the toolbar & keyboard shortcuts for "Build," "Run,", "Build & Run," etc. to it, no matter which file is being edited.

For example, when making lots of changes to the game, you will generally have the "endless-sky-lib" project active, so that you can ensure changes compile without needing to fully link things together into the actual game binary. After making changes, you would then activate the "EndlessSky" project and use the "Build & Run" option with the "Release" target, which will fully link the compiled code into the small, performant executable.

You can continue making tweaks to files that belong to the "endless-sky-lib" project without reactivating it, so that "Build & Run" will still launch the game with your latest tweaks. Should you experience some show-stopper bug in your modified game, you would then change to the "Debug" build target to help determine the source(s) of the issue.

Before the game will run the first time, however, you will need to skip down to the instructions at **Running the game** below.

> ### Building with Scons

Reviewer: Scons section should come first, as the official build.

If you instead want to build the game from the command line via Scons, you will need [Python 3.8 or later](https://www.python.org/downloads/). When installing, be sure to select the "Add to PATH" checkbox.

Afterwards, you will need to add your MinGW installation to your path manually. To do so on Windows 10, to to **Settings** > **System** > **About** > **System info** > **Advanced system settings** > **Environment Variables**. From each, select **Path** under System variables, and click Edit.

Click New, and then if you used a default MinGW installation, enter `C:\Program Files\mingw-w64\x86_64-8.1.0-posix-seh-rt_v6-rev0\mingw64\bin` as the path, then hit OK to exit out of these windows.

> From there, restart your command line if you had one open, and simply `cd` to your Endless Sky directory, then type `mingw32-make.exe -f .winmake`. To speed up the build process, you can add a `-jX` flag to that command, where X is the number of CORES/THREADS? your processor has.

Reviewer: ```suggestion
From there, restart your command line if you had one open, and simply `cd` to your Endless Sky directory, then type `mingw32-make.exe -f .winmake`. (`make`'s standard `-j N` flag is supported, to increase the number of active parallel build tasks to N.)
```

Author: What should a user put in for N? Is it based on threads or cores? I'm guessing threads, but just want to be sure.

Reviewer: Neither, as with all parallelism the choice of "how many jobs" is dependent on both the system and the amount of work to be done.

Author: Can we offer a guideline? This is meant for new developers, and the workload is relatively fixed. It doesn't take more than 5 minutes on my terrible laptop. So using a quad-core processor as an example, like an i5-6600k, how many jobs would you run?

Reviewer 2: the most common use case is maxing out your CPU, which is achieved by N = Number of logical cores. You may need to clarify this use case to make tehh happy.

Reviewer: We shouldnt need to provide any guideline on the value. Explaining how to invoke parallel builds should be enough to get developers started, either learning about "make" and the `-j` flag (because they're new to compiling in general) or just to building (e.g. because they're new to `scons` and didn't know it supports parallel builds, too), or experimenting with different values (because they're curious about the possible speedups).

As the intended audience of this document is people familiar with building software programs, or at minimum intent on learning to build this program, I think it is not our place to explain how other programs work - those programs have entire websites devoted to them with that purpose!

Author: We don't _need_ to, sure, but the whole point of a guide is to provide guidance. If MCO is right, and the most common use case is to put in your CPU's logical processors for the number of tasks, then I see no harm mentioning that. If not, and you really need to do lots of in depth reading or experimentation to find anything close to useful, then it's fine as-is.

Reviewer 3: As a new developer for whom scons is a misspelling of a tasty bread product rather than something to do with games, having the guide spell out common flags and how to use them would be appreciated. It is a guide, after all, and having it be all-inclusive as to what it takes for me to build things quickly and effectively would be much appreciated.

The main thing, I suppose, is to not make any assumptions about our developer's knowledge.

> ### Building with a third-party IDE

Reviewer: ```suggestion
### Building with other IDEs
```

Refer to the build pipeline files under `.github\` to tailor the build process to your chosen IDE.

### Running the Game

For Endless Sky to run on Windows, it needs various .dll files included in the base directory. Go to `C:\dev64\bin\`, and copy the files to your base directory. These should include:

  - glew32.dll
  - **libgcc_s_seh-1.dll**
  - libjpeg-62.dll
  - libmad-0.dll
  - libpng15-15.dll
  - **libstdc++-6.dll**
  - libturbojpeg.dll
  - libwinpthread-1.dll
  - OpenAL32.dll
  - SDL2.dll
  - soft_oal.dll
  - zlib1.dll

For the **bolded** libraries, those will need to be included from your MinGW installation, instead. By default, you can go to `C:\Program Files\mingw-w64\x86_64-8.1.0-posix-seh-rt_v6-rev0\mingw64\bin`, and copy those libraries to your Endless Sky directory to overwrite the current ones.

> **TODO:** Incorporate these libraries into the provided zip, so the user only has to copy from one source.

Reviewer: This won't be done

Author: The only reason mentioned on Discord was we couldn't be sure what version of MinGW/g++ the user was using. Now, we specify for new developers to use 8.1.0. And even in that case, I went back and test 8.1.0, 7.1.0, 6.3.0, and 5.1.0. All compiled without a hitch on both my current installation and on a VM, and the game ran with only needing to copy from one directory. Is there some other reason I'm missing?
